python/mujoco_mpc/demos/predictive_sampling/predictive_sampling.py

```diff
@@ -0,0 +1,331 @@
+    return 0.5 * (output[:, bounds[1]] - output[:, bounds[0]]) / (
+        input[bounds[1]] - input[bounds[0]]
+    ) + 0.5 * (output[:, bounds[0]] - output[:, bounds[0] - 1]) / (
+        input[bounds[0]] - input[bounds[0] - 1]
+    )
+
+  # get action from policy
+  def action(self, time: float) -> np.array:
+    # find interval containing time
+    bounds = self._find_interval(self._times, time)
+
+    # boundary case
+    if bounds[0] == bounds[1]:
+      return self.clamp(self._parameters[:, bounds[0]])
+
+    # normalized time
+    t = (time - self._times[bounds[0]]) / (
+        self._times[bounds[1]] - self._times[bounds[0]]
+    )
+
+    if self._interp == "cubic":
+      # spline coefficients
+      c0 = 2.0 * t * t * t - 3.0 * t * t + 1.0
+      c1 = (t * t * t - 2.0 * t * t + t) * (
+          self._times[bounds[1]] - self._times[bounds[0]]
+      )
+      c2 = -2.0 * t * t * t + 3 * t * t
+      c3 = (t * t * t - t * t) * (
+          self._times[bounds[1]] - self._times[bounds[0]]
+      )
+
+      # slopes
+      m0 = self._slope(self._times, self._parameters, self._times[bounds[0]])
+      m1 = self._slope(self._times, self._parameters, self._times[bounds[1]])
+
+      # interpolation
+      return self.clamp(
+          c0 * self._parameters[:, bounds[0]]
+          + c1 * m0
+          + c2 * self._parameters[:, bounds[1]]
+          + c3 * m1
+      )
+    elif self._interp == "linear":
+      return self.clamp(
+          (1.0 - t) * self._parameters[:, bounds[0]]
+          + t * self._parameters[:, bounds[1]]
+      )
+    else:  # self._interp == "zero"
+      return self.clamp(self._parameters[:, bounds[0]])
+
+  # resample policy plan from current time
+  def resample(self, time: float):
+    # new times and parameters
+    times = np.array(
+        [i * self._splinestep + time for i in range(self._nspline)], dtype=float
+    )
+    parameters = np.vstack([self.action(t) for t in times]).T
+
+    # update
+    self._times = times
+    self._parameters = parameters
+
+  # add zero-mean Gaussian noise to policy parameters
+  def add_noise(self, scale: float):
+    # clamp within limits
+    self._parameters = self.clamp(
+        self._parameters
+        + np.random.normal(scale=scale, size=(self._naction, self._nspline))
+    )
+
+  # return a copy of the policy with noisy parameters
+  def noisy_copy(self, scale: float) -> Policy:
+    # create new policy object
+    policy = Policy(self._naction, self._horizon, self._splinestep)
+
+    # copy policy parameters into new object
+    policy._parameters = np.copy(self._parameters)
+
+    # get noisy parameters
+    policy.add_noise(scale)
+
+    return policy
+
+  # clamp action with limits
+  def clamp(self, action: np.array) -> np.array:
+    # clamp within limits
+    if self._limits is not None:
+      return np.minimum(
+          np.maximum(self._limits[:, 0], action), self._limits[:, 1]
+      )
+    return action
+
+
+# rollout
+def rollout(
```

Comment:
Potentially replace with MuJoCo threaded Python rollout function.

Reply:
I'll add a TODO.  Once this is put to use and performance becomes an issue, this change would be a ripe low-hanging fruit.